test/integration/join_cluster_test.js

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var _ = require('underscore');
+var after = require('after');
+var DebuglogLogger = require('../../lib/debug-log-logger.js');
+var Ringpop = require('../../index.js');
+var TChannel = require('tchannel');
+var test = require('tape');
+
+function bootstrapClusterOf(size, opts, tap, onBootstrap) {
+    var cluster = createClusterOf(size, opts);
+
+    var bootstrapHosts = cluster.map(function mapRingpop(ringpop) {
+        return ringpop.hostPort;
+    });
+
+    if (typeof tap === 'function') {
+        tap(cluster);
+    }
+
+    for (var i = 0; i < cluster.length; i++) {
+        cluster[i].bootstrap(bootstrapHosts, Array.isArray(onBootstrap) ?
+            onBootstrap[i] : onBootstrap);
+    }
+
+    return cluster;
+}
+
+function createClusterOf(size, opts) {
+    size = size || 3;
+
+    var cluster = [];
+
+    for (var i = 0; i < size; i++) {
+        cluster.push(createRingpop(_.extend({
+            host: '127.0.0.1',
+            port: 10000 + i
+        }, opts)));
+    }
+
+    return cluster;
+}
+
+function createTChannel(host, port) {
+    return new TChannel({
+        host: host,
+        port: port
+    });
+}
+
+function createRingpop(opts) {
+    opts = opts || {};
+
+    var ringpop = new Ringpop(_.extend({
+        app: 'test',
+        hostPort: opts.host + ':' + opts.port,
+        maxJoinDuration: opts.maxJoinDuration,
+        channel: createTChannel(opts.host, opts.port),
+        logger: DebuglogLogger()
+    }, opts));
+
+    ringpop.setupChannel();
+
+    return ringpop;
+}
+
+function destroyCluster(cluster) {
+    cluster.forEach(function eachRingpop(ringpop) {
+        ringpop.destroy();
+    });
+}
+
+test('one node can join', function t(assert) {
```

Comment:
suggestion: consider having a `bootstrapCluster.test` helper.

```js
bootstrapCluster.test = function test(name, opts, fn) {
    test(name, function onAssert(assert) {
        assert.once('end', onEnd);

        var cluster = bootstrapClusterOf(
            opts.size, opts, opts.tap, onBootstrap
        );

        function onEnd() {
            destroyCluster(cluster);
        }

        function onBootstrap(err) {
            if (err) {
                assert.ifError(err);
                return assert.end();
            }

            fn(cluster, assert);
        }
    });
};


bootstrapCluster.test('one node can join', {
    size: 1
}, function t(cluster, assert) {
    assert.end();
});
```

Thank @jcorbin for this pattern.